zapisy/apps/schedule/models/term.py

```diff
@@ -210,59 +210,87 @@ def __str__(self):
+    rooms = course_term.classrooms.all() if course_term.pk else None
     for day in dates:
-        rooms = instance.classrooms.all()
-        for room in rooms:
+        if rooms:
+            for room in rooms:
+                Term.objects.create(event=event,
+                                    day=day,
+                                    start=course_term.start_time,
+                                    end=course_term.end_time,
+                                    room=room)
+        else:
             Term.objects.create(event=event,
                                 day=day,
-                                start=instance.start_time,
-                                end=instance.end_time,
-                                room=room)
-        if not rooms:
-            Term.objects.create(event=event,
-                                day=day,
-                                start=instance.start_time,
-                                end=instance.end_time,
+                                start=course_term.start_time,
+                                end=course_term.end_time,
                                 room=None)
+
+
+"""
+In the functions below instance_db is the version of a CourseTerm found in
+the database and instance_curr is its most recent version.
+We want to keep the following invariant:
+All fields of Terms associated with a CourseTerm correspond to the fields of
+the CourseTerm's version currently stored in the database.
+"""
+
+
+@receiver(models.signals.pre_delete, sender=CourseTerm)
+def sync_course_term_delete(**kwargs):
+    """Deletes all Terms associated with a CourseTerm when it is deleted."""
+    instance_curr: CourseTerm = kwargs['instance']
+    if instance_curr.pk:
+        instance_db: CourseTerm = CourseTerm.objects.get(pk=instance_curr.pk)
+        delete_terms(instance_db)
+
+
+@receiver(models.signals.pre_save, sender=CourseTerm)
+def sync_course_term_save(**kwargs):
+    """Creates or updates Terms associated with a CourseTerm when it is saved.
+
+    If the CourseTerm is already in the database, all Terms associated with it
+    are deleted and new ones (with updated fields) are created.
+    """
+    instance_curr: CourseTerm = kwargs['instance']
+    if instance_curr.pk:
+        instance_db: CourseTerm = CourseTerm.objects.get(pk=instance_curr.pk)
+        delete_terms(instance_db)
+    create_terms(instance_curr)
+
+
+@receiver(models.signals.m2m_changed, sender=CourseTerm.classrooms.through)
+def sync_course_term_m2m(**kwargs):
+    """Creates or deletes Terms associated with a CourseTerm when its set of classrooms changes.
+
+    The Terms associated with the CourseTerm are simply deleted and recreated.
+    If a classroom has been removed from the set, Terms having its id in their
+    'room' field will not be recreated. If a classroom has been added to the
+    set, new Terms having its id in their 'room' field will be created.
```

Comment:
Można to napisać nieco bardziej w języku naturalnym, np. jakoś tak:
```suggestion
    If a classroom has been removed from the set, Terms taking place in it
    will not be recreated. If a classroom has been added to the set, new Terms
    taking place in it will be created.
```